src/EventStore.Client.Streams/EventStoreClient.Append.cs

```diff
@@ -189,5 +193,130 @@ public partial class EventStoreClient {
 
 			return writeResult;
 		}
+
+#if NETCOREAPP3_1_OR_GREATER
+		private class StreamAppender : IDisposable {
+			private readonly EventStoreClientSettings _settings;
+			private readonly CancellationToken _cancellationToken;
+			private readonly Action<Exception> _onException;
+			private readonly Channel<BatchAppendReq> _channel;
+			private readonly ConcurrentDictionary<Uuid, TaskCompletionSource<IWriteResult>> _pendingRequests;
+
+			private readonly Task<AsyncDuplexStreamingCall<BatchAppendReq, BatchAppendResp>> _callTask;
+
+			public StreamAppender(EventStoreClientSettings settings,
+				Task<AsyncDuplexStreamingCall<BatchAppendReq, BatchAppendResp>> callTask, CancellationToken cancellationToken,
+				Action<Exception> onException) {
+				_settings = settings;
+				_callTask = callTask;
+				_cancellationToken = cancellationToken;
+				_onException = onException;
+				_channel = System.Threading.Channels.Channel.CreateBounded<BatchAppendReq>(10000);
+				_pendingRequests = new ConcurrentDictionary<Uuid, TaskCompletionSource<IWriteResult>>();
+				_ = Task.Factory.StartNew(Send, TaskCreationOptions.LongRunning);
+				_ = Task.Factory.StartNew(Receive, TaskCreationOptions.LongRunning);
+			}
+
+			public ValueTask<IWriteResult> Append(string streamName, StreamRevision expectedStreamPosition,
+				IEnumerable<EventData> events, TimeSpan? timeoutAfter, CancellationToken cancellationToken = default) =>
+				AppendInternal(BatchAppendReq.Types.Options.Create(streamName, expectedStreamPosition, timeoutAfter),
+					events, cancellationToken);
+
+			public ValueTask<IWriteResult> Append(string streamName, StreamState expectedStreamState,
+				IEnumerable<EventData> events, TimeSpan? timeoutAfter, CancellationToken cancellationToken = default) =>
+				AppendInternal(BatchAppendReq.Types.Options.Create(streamName, expectedStreamState, timeoutAfter),
+					events, cancellationToken);
+
+			private async Task Receive() {
+				var call = await _callTask.ConfigureAwait(false);
+				try {
+					await foreach (var response in call.ResponseStream.ReadAllAsync(_cancellationToken)
+						.ConfigureAwait(false)) {
+						if (!_pendingRequests.TryRemove(Uuid.FromDto(response.CorrelationId), out var writeResult)) {
+							continue; // TODO: Log?
```

Comment:
Do we want to decide on whether to log this before merging?

Reply:
we shouldn't log, if it has already been removed then the result is what ever was received and and any subsequent message would be from a race condition and should be ignored